discovery/sync_manager.go

```diff
@@ -0,0 +1,673 @@
+
+		s = m.chooseRandomSyncer(candidatesChosen, true)
+	}
+}
+
+// chooseRandomSyncer returns a random non-active syncer that's eligible for a
+// sync transition. A blacklist can be used to skip any previously chosen
+// candidates. The useActive boolean can be used to also filter active syncers.
+//
+// NOTE: It's possible for a nil value to be returned if there are no eligible
+// candidate syncers.
+//
+// NOTE: This method must be called with the syncersMtx lock held.
+func (m *SyncManager) chooseRandomSyncer(blacklist map[routing.Vertex]struct{},
+	useActive bool) *GossipSyncer {
+
+	eligible := func(s *GossipSyncer) bool {
+		// Skip any syncers that exist within the blacklist.
+		if blacklist != nil {
+			if _, ok := blacklist[s.cfg.peerPub]; ok {
+				return false
+			}
+		}
+
+		// Only syncers in a chansSynced state are viable for sync
+		// transitions, so skip any that aren't.
+		return s.syncState() == chansSynced
+	}
+
+	for _, s := range m.inactiveSyncers {
+		if !eligible(s) {
+			continue
+		}
+		return s
+	}
+
+	if useActive {
+		for _, s := range m.activeSyncers {
+			if !eligible(s) {
+				continue
+			}
+			return s
+		}
+	}
+
+	return nil
+}
+
+// InitSyncState is called by outside sub-systems when a connection is
+// established to a new peer that understands how to perform channel range
+// queries. We'll allocate a new gossip syncer for it, and start any goroutines
+// needed to handle new queries.
+//
+// TODO(wilmer): Only mark as ActiveSync if this isn't a channel peer.
+func (m *SyncManager) InitSyncState(peer lnpeer.Peer) {
+	// If we already have a syncer, then we'll exit early as we don't want
+	// to override it.
+	nodeID := routing.Vertex(peer.PubKey())
+	if _, ok := m.GossipSyncer(nodeID); ok {
+		return
+	}
+
+	log.Infof("Creating new GossipSyncer for peer=%x", nodeID[:])
+
+	encoding := lnwire.EncodingSortedPlain
+	s := newGossipSyncer(gossipSyncerCfg{
+		chainHash:     m.cfg.ChainHash,
+		peerPub:       nodeID,
+		channelSeries: m.cfg.ChanSeries,
+		encodingType:  encoding,
+		chunkSize:     encodingTypeToChunkSize[encoding],
+		sendToPeer: func(msgs ...lnwire.Message) error {
+			return peer.SendMessage(false, msgs...)
+		},
+	})
+
+	// Gossip syncers are initialized by default as passive and in a
+	// chansSynced state so that they can reply to any peer queries or
+	// handle any sync transitions.
+	s.setSyncType(PassiveSync)
+	s.setSyncState(chansSynced)
+	s.Start()
+
+	m.Lock()
+	m.inactiveSyncers[nodeID] = s
+	numActiveSyncers := len(m.activeSyncers) + len(m.pendingActiveSyncers)
+	m.Unlock()
+
+	switch {
+	// If we don't have any active syncers, we'll force a historical sync
+	// with the first peer we connect to, to ensure we get as much of the
+	// graph as possible. This is mostly done for light clients, which
+	// shouldn't be receiving new updates anyway due to the high bandwidth
+	// consumption that stems from it.
+	case m.cfg.NumActiveSyncers == 0:
```

Comment:
imo this should be it's own config option, idt it's logical that if you set the number of active syncers to zero that it will go ahead and do a historical sync. i can also imagine cases where you want both

Reply:
Fixed.